### ISSUE TITLE
gRPC cluster configuration

### ISSUE THREAD

**Issue by vadimi** [turbinelabs/rotor#7]
Hello,

I have some gRPC services that I'd like to proxy through envoy and rotor. Is there a way to specify [Http2ProtocolOptions](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/protocol.proto#core-http2protocoloptions) on a cluster?

**Comment by 9len** [comment]
We are working on a method to staticly configure clusters and listeners at rotor startup. Stay tuned!

**Comment by vadimi** [comment]
I was hoping for something dynamic, as my services are in DCOS and ECS. Maybe additional labels like `grpc: true` would work? Something like they have in [Ambassador grpc integration](https://www.getambassador.io/user-guide/grpc)

**Comment by 9len** [comment]
Do you add and remove services regularly, or is the dynamism mostly about service discovery? If the latter, static listener configs would work fine. 

The other thing we're considering is being able to specify a new listener _template_, so each new listener added would use the same template.

The reason we'd rather stick to either statically defined listeners or listener templates is that we'd like to avoid reinventing the surface area of envoy configuration; the existing envoy configuration schema was purpose built to describe this sort of stuff.

Fully dynamic LDS/RDS is unfortunately out of scope for Rotor, since it requires a durable backing-store; that’s part of what we sell as our commercial product, [Houston](turbinelabs.io/product). However, Envoy makes it reasonable to separate your CDS/EDS (service discovery) implementation from your LDS/RDS (routing) implementation, and this is our suggestion for anyone who needs real dynamism at the routing layer.

**Comment by 9len** [comment]
gah! reading this again, I now understand you want to set this for a cluster. 

If the names of the clusters themselves change fairly rarely, and you don't add them all that often, then a statically configured cluster will work fine; the instances can still be dynamically collected.

If all of your clusters have the same configuration, a new-cluster template will work. Some combination of a template and specific static overrides would probably work ok.

**Comment by vadimi** [comment]
We have a few gRPC services and a few HTTP services in one ECS cluster. This is why I need to be able to dynamically specify protocol based on the service. The names change fairly frequently as we test multiple feature branches in parallel. Statically configured listeners should be fine, but picking `http2_protocol_options` or `http1_protocol_options` (or make it a default) for envoy cluster per service is definitely something dynamic in my mind.

**Comment by 9len** [comment]
Thanks for the context. 

We will continue to ponder this, but for the moment, dynamic cluster/route/listener configuration is not a core goal for the project.

One approach that might work, assuming we ship a "new cluster template" option, would be to set up a pair of Rotors, each collecting a different label, one for HTTP services, one for gRPC. Then point your envoys at both rotors in the static resources. I _think_ that will give you the dynamism you need.

**Comment by 9len** [comment]
static cluster/listener configuration landed in 7acd7cd08a3512d0424a6b742033abb5d50f295c, available in the [`0.18.0`](https://github.com/turbinelabs/rotor/commits/0.18.0) tag.
